### DISCUSSION TITLE
“docker run -d --name dind-rootless --privileged docker:27.3.1-dind-rootless” does not support cgroupfs driver

### DISCUSSION THREAD

**Discussion by yushiqie** [moby/moby#48817] [Q&A]
Rootless-mode cgroup driver only supports systemd, why doesn't it support cgroupfs?

The docker info output information(in contianer dind-rootless) is as follows:
![20241105165534](https://github.com/user-attachments/assets/69ddc598-1310-45e0-b037-e530a6797281)

host os info：
![20241105171241](https://github.com/user-attachments/assets/018ba1e0-39e7-4d0a-974b-e5d1e409a56d)

**Comment by AkihiroSuda** [comment]
> Rootless-mode cgroup driver only supports systemd, why doesn't it support cgroupfs?

Because there is no easy way to delegate cgroup v2 without systemd

**Comment by yushiqie** [comment]
I'm a bit confused, when running dind-rootless as the root user, it supports the cgroupfs driver. What is the reason for the difference in support for the cgroupfs driver between the root and non-root users inside the container？

`docker run -d --name dind-rootless --user root --privileged docker:27.3.1-dind-rootless`

![20241114112118](https://github.com/user-attachments/assets/37062d8a-1fb6-4b94-83bc-1d0b54fc0c72)

**Reply by AkihiroSuda** [reply]
The difference is that /sys/fs/cgroup/.../CONTAINERID has to be chowned by something (e.g. systemd) for rootless 

Maybe we can support rootless cgroupfs if we can extend the parent docker to chown it for rootless

**Reply by AkihiroSuda** [reply]
BTW dind-rootless image is not expected to work with --user root.
I'm surprised to know that it works

**Reply by yushiqie** [reply]
> /sys/fs/cgroup/.../CONTAINERID
If user `rootless` has permissions on `/sys/fs/cgroup/.../CONTAINERID`, then the cgroupfs driver can be supported ? Currently, /sys/fs/cgroup is read-write accessible to the `rootless`, and the `rootless` can create new directories such as /sys/fs/cgroup/test.

![20241114163937](https://github.com/user-attachments/assets/17dff8c6-a0e0-4d5f-9a2d-af781e475cba)

**Reply by yushiqie** [reply]
> BTW dind-rootless image is not expected to work with --user root. I'm surprised to know that it works

Just a beautiful mistask，but it works well 😁

**Reply by AkihiroSuda** [reply]
> If user rootless has permissions on /sys/fs/cgroup/.../CONTAINERID, then the cgroupfs driver can be supported ? 

Yes, it should be easy to implement it.

> Currently, /sys/fs/cgroup is read-write accessible to the rootless, and the rootless can create new directories such as /sys/fs/cgroup/test.

Yes, I have completely forgotten that this was already implemented three years ago 😅 
- https://github.com/opencontainers/runtime-spec/pull/1123
- https://github.com/opencontainers/runc/pull/3057

**Reply by yushiqie** [reply]
> > If user rootless has permissions on /sys/fs/cgroup/.../CONTAINERID, then the cgroupfs driver can be supported ?
> 
> Yes, it should be easy to implement it.

Great! In which version Docker will implement this feature?

**Reply by yushiqie** [reply]
> Yes, I have completely forgotten that this was already implemented three years ago 😅
> 
> * [specify cgroup ownership semantics opencontainers/runtime-spec#1123](https://github.com/opencontainers/runtime-spec/pull/1123)
> * [chown cgroup to process uid in container namespace opencontainers/runc#3057](https://github.com/opencontainers/runc/pull/3057)

I would like to learn more about which module in cgroupv2 chown cgroup to rootless user, and specifically where is implemented in the Docker code. Additionally, if cgroupv2 can chown cgroup to  rootless user, can the same be achieved with cgroupv1 ?

**Reply by AkihiroSuda** [reply]
>  In which version Docker will implement this feature?

The first major version right after somebody submits a PR

> I would like to learn more about which module in cgroupv2 chown cgroup to rootless user, and specifically where is implemented in the Docker code.

Chowned by runc (see the link above)

> Additionally, if cgroupv2 can chown cgroup to rootless user, can the same be achieved with cgroupv1 ?

Not considered to be safe
https://github.com/systemd/systemd/blob/a2240d1cf9f3515728186c2c98a6f4b64a40e4da/docs/CGROUP_DELEGATION.md?plain=1#L228-L230
